### ISSUE TITLE
ER: distinct(s) - an efficient, stream-oriented counterpart to `unique/0` 

### ISSUE THREAD

**Issue by pkoppstein** [stedolan/jq#1555]
The existing `unique` builtin currently sorts its input, and only works on arrays.  Since it has been in jq for a long while, it is unlikely to change in either respect.  Hence this proposal.  A reference implementation would be:

    def distinct(s): reduce s as $x ({}; .[$x | (type[0:1] + tostring)] = $x) |.[];

Note that this is intended to exploit jq's optimization of key lookups without risking collisions.

One important use-case for such a function is in connection with jq's support for "relational" (SQLish) functionality: if the jq builtins ensure uniqueness, then they could use distinct(s); otherwise, users may want to do so.

An alternative implementation approach would be to define the dictionary-creation function separately:

    def dict(s): reduce s as $x ({}; .[$x | (type[0:1] + tostring)] = $x);
    def distinct(s): dict(s)[];

If `dict/1` or equivalent is included in builtin.jq, then it would be appropriate to make it easy to check whether an entity is already in the dictionary, e.g. 

    # input: a dictionary using `type[0:1] + tostring`-style keys
    def lookup($item): .[$item | (type[0:1] + tostring)];

**Comment by nicowilliams** [comment]
Nice.

Let's not have `dict(s)` but `set(s)`.  When we as coroutines we can add a proper dict.

**Comment by pkoppstein** [comment]
As previously mentioned, I'm not in a position to offer a PR, but here are some pertinent bits and pieces.

### builtin.jq
```
# emit a set-representation of items in the stream, s
def set(s): reduce s as $x ({}; .[$x | (type[0:1] + tostring)] = $x);

# emit a stream of the distinct entities in the stream s, without sorting them
def distinct(s): set(s)[];

# input: a dictionary using keys as produced by set/1
def lookup($item): .[$item | (type[0:1] + tostring)];  
```
### jq.test

[distinct(1,1,"1", [1,1],[1,1])]
null
[1,"1",[1,1]]

set({a:1}, {a:1}, {a:2}) | lookup({a:1})
null
{"a":1}

**Comment by nicowilliams** [comment]
Yes, that works.  For a proper `dict`, what I want is to add co-routines then do something like:

```
def dict(keys; values):
    def pairs(keys; values):
        def pairs: (.,[first(keys),first(values)])|select(length == 2)|pairs; pairs;
    reduce (pairs(@@keys; @@values)) as .[$key, $value] ({}; .[$key|(type[0:1] + tostring)] = $value);
```

Then your `lookup` can be used for sets and dicts.

**Comment by fadado** [comment]
Do not forget infinite streams:

```
def distinct(stream):
    foreach stream as $x (
        {};
        ($x|type[0:1]+tostring) as $k
            | if has($k)
              then .["~"]=false
              else .[$k]=$x | .["~"]=true end;
        select(.["~"]) | $x)
;
```

**Comment by pkoppstein** [comment]
@fadado - `distinct(s)` does not make any sense for infinite streams as one can only be sure that an item is unique in the stream after the entire stream has been seen; by contrast, `nondistinct(s)` would make sense for such streams, though I'm wondering how useful it would be.  

(Of course `uniq(s)` in the sense of eliminating adjacent repetitions would also make sense for infinite streams.)

**Comment by odnoletkov** [comment]
I think `distinct` makes perfect sense for infinite streams just like `uniq` does. It should just remove duplicates and this could be done in a streaming fashion. Of course it would not work well for streams with mostly distinct items. But otherwise it could be very efficient and useful.

**Comment by pkoppstein** [comment]
@odnoletkov - Please explain the semantics you have in mind for `stream(s)` for an infinite stream, s. Note that the issue here is not for finite streams of indeterminate length but for streams that are conceptually infinite (i.e. do not terminate, e.g. `1|recurse(.)`).

**Comment by leonid-s-usov** [comment]
@pkoppstein I think that @odnoletkov is right and there shouldn't be any restriction on using the distinct filter even with infinite streams. It would be a way to shoot oneself into the leg, but just as wasteful would it be to attempt build an array with a "conceptually infinite stream"
truly infinite streams are in any case quite specific tool, and one using those will probably have some clear plan on how to use it and whether one or another filter would apply.

**Comment by pkoppstein** [comment]
@leonid-s-usov - You seem also to have in mind some specific semantics for `distinct(s)` that differ from the intended semantics in this thread.  

The point is that jq can only examine the items in a stream in sequence, and therefore `distinct(s)` can only emit an item once the entire stream has been seen, and therefore given a truly unbounded stream, `distinct(s)` would always just "hang".  That is, it cannot emit ANYTHING until the "end-of-stream" has been encountered.  

Perhaps it will help to remind you that jq could not "sort" an infinite stream.

`uniq` is quite different, because once a transition has been made, a value can be emitted.

**Comment by odnoletkov** [comment]
To illustrate my point on the infinite streams

    $ jq -n 'distinct(0 | recurse(. + 1))' | head -5

I believe @fadado version would would produce expected results and won't hang

**Comment by pkoppstein** [comment]
@fadado - Good point!

@odnoletkov, @leonid-s-usov - Sorry for the mixup on my part.
